R-package/tests/testthat/test_Predictor.R

```diff
@@ -111,3 +113,132 @@ test_that("start_iteration works correctly", {
     pred_leaf2 <- predict(bst, test$data, start_iteration = 0L, num_iteration = end_iter + 1L, predleaf = TRUE)
     expect_equal(pred_leaf1, pred_leaf2)
 })
+
+.expect_has_row_names <- function(pred, X) {
+    if (is.vector(pred)) {
+        rnames <- names(pred)
+    } else {
+        rnames <- row.names(pred)
+    }
+    expect_false(is.null(rnames))
+    expect_true(is.vector(rnames))
+    expect_equal(row.names(X), rnames)
+}
+
+.expect_doesnt_have_row_names <- function(pred) {
+    if (is.vector(pred)) {
+        expect_null(names(pred))
+    } else {
+        expect_null(row.names(pred))
+    }
+}
+
+.expect_row_names_kept <- function(bst, X, multiclass = FALSE) {
+    pred <- predict(bst, X, reshape = TRUE)
+    .expect_has_row_names(pred, X)
+    pred <- predict(bst, X, reshape = TRUE, rawscore = TRUE)
+    .expect_has_row_names(pred, X)
+    pred <- predict(bst, X, reshape = TRUE, predleaf = TRUE)
+    .expect_has_row_names(pred, X)
+    pred <- predict(bst, X, reshape = TRUE, predcontrib = TRUE)
+    .expect_has_row_names(pred, X)
+    pred <- predict(bst, X, reshape = FALSE)
+    if (!multiclass || NROW(X) == NROW(pred)) {
+        .expect_has_row_names(pred, X)
+    } else {
+        .expect_doesnt_have_row_names(pred)
+    }
```

Comment:
I'm really struggling to understand these test cases, since it requires reasoning about what was passed in to `.expect_row_names_kept()` from within each of the `test_that(...)` calls, and since these `if` statements rely on the fact that `pred` here happens to reference the predictions produced with `reshape = FALSE` and none of `rawscore` / `predleaf` / `predcontrib`.

To make this easier to follow (both for reviewing and for future contributors), can you please remove the use of this `.expect_row_names_kept()`  function and instead explicitly include checks in each of the `test_that(...)` calls? Those checks should not need to use `NROW(pred)` at all. For test code, I'd prefer some code duplication in exchange for it being easier to understand what is being tested.

Reply:
Not sure what you want me to do there. That `pred` *does* reference predictions with different combinations of `reshape` / `rawscore` / `predcontrib` / `predleaf` (all predictions generated there are named `pred`).

You asked me to add tests for different values of `reshape`, but `reshape` might get ignored depending on the rest of the parameters, and thus whether the output is meant to have row names or not, depends on the shape of what's actually returned in the end by the prediction function. If I were to adjust the tests by hard-coding the reshape logic in them, they then would test both the row names and the reshape logic together, which I don't think is ideal. The most logical way of testing the row names in a way that's independent of `reshape` would be by checking the number of rows in the data and the predictions.

Reply:
> they then would test both the row names and the reshape logic together, which I don't think is ideal

"hard-coding the reshape logic", i.e. not changing the test expectation based on a condition like ` NROW(X) == NROW(pred)`, is exactly what I'm asking for.

We know, for example, that for `reshape = FALSE` and a multi-class objective, the output of `predict()` is expected to have `NROW(X) * number_of_classes` elements, and therefore shouldn't have row names. It's desirable for the tests to break if that changes.

As a maintainer on this project, I believe I'm able to push commits onto branches from forks associated with open PRs. Would you like me to just directly push the testing changes I'm proposing to this branch?

Reply:
Yes, would be better if you modify the tests yourself.

Reply:
Ok no prob! I'll push a commit in the next day or two.

Reply:
Updated in https://github.com/microsoft/LightGBM/pull/4977/commits/04b743fd94e6d398909209adf306275ab67816fd.

Thankfully, #4971  helped simplify these tests (and therefore users' code). Thanks for that!